src/minimap.cpp

```diff
@@ -0,0 +1,475 @@
+/*
+Minetest
+Copyright (C) 2010-2015 celeron55, Perttu Ahola <celeron55@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "minimap.h"
+#include "logoutputbuffer.h"
+#include "jthread/jmutexautolock.h"
+#include "clientmap.h"
+#include "settings.h"
+#include "nodedef.h"
+#include "porting.h"
+#include "util/numeric.h"
+#include "util/string.h"
+#include <math.h>
+
+
+
+QueuedMinimapUpdate::QueuedMinimapUpdate():
+	p(-1337,-1337,-1337),
+	data(NULL)
+{
+}
+
+QueuedMinimapUpdate::~QueuedMinimapUpdate()
+{
+	delete data;
+}
+
+MinimapUpdateQueue::MinimapUpdateQueue()
+{
+}
+
+MinimapUpdateQueue::~MinimapUpdateQueue()
+{
+	JMutexAutoLock lock(m_mutex);
+
+	for(std::vector<QueuedMinimapUpdate*>::iterator
+			i = m_queue.begin();
+			i != m_queue.end(); i++)
+	{
+		QueuedMinimapUpdate *q = *i;
+		delete q;
+	}
+}
+
+void MinimapUpdateQueue::addBlock(v3s16 p, MinimapMapblock *data)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	JMutexAutoLock lock(m_mutex);
+
+	/*
+		Find if block is already in queue.
+		If it is, update the data and quit.
+	*/
+	for(std::vector<QueuedMinimapUpdate*>::iterator
+			i = m_queue.begin();
+			i != m_queue.end(); i++)
+	{
+		QueuedMinimapUpdate *q = *i;
+		if(q->p == p) {
+			delete q->data;
+			q->data = data;
+			return;
+		}
+	}
+
+	/*
+		Add the block
+	*/
+	QueuedMinimapUpdate *q = new QueuedMinimapUpdate;
+	q->p = p;
+	q->data = data;
+	m_queue.push_back(q);
+}
+
+QueuedMinimapUpdate * MinimapUpdateQueue::pop()
+{
+	JMutexAutoLock lock(m_mutex);
+
+	for(std::vector<QueuedMinimapUpdate*>::iterator
+			i = m_queue.begin();
+			i != m_queue.end(); i++)
+	{
+		QueuedMinimapUpdate *q = *i;
+		m_queue.erase(i);
+		return q;
+	}
+	return NULL;
+}
+
+/*
+	Minimap update thread
+*/
+
+void * MinimapUpdateThread::Thread()
+{
+	ThreadStarted();
+
+	log_register_thread("MinimapUpdateThread");
+
+	DSTACK(__FUNCTION_NAME);
+
+	BEGIN_DEBUG_EXCEPTION_HANDLER
+
+	porting::setThreadName("MinimapUpdateThread");
+
+	while(!StopRequested()) {
+
+		while(m_queue.size()) {
+			QueuedMinimapUpdate *q = m_queue.pop();
+			std::map<v3s16, MinimapMapblock *>::iterator it;
+			it = m_blocks_cache.find(q->p);
+			if (q->data)
+				m_blocks_cache[q->p] = q->data;
+			else if (it != m_blocks_cache.end())
```

Comment:
This piece not insterts new block, modifies stored one or removes it from cache if data is empty.